### ISSUE TITLE
[2.0] Classes which are generics should document their type parameters

### ISSUE THREAD

**Issue by sirosen** [sqlalchemy/sqlalchemy#9078]
### Describe the use case

I'm working on a test branch for sqlalchemy 2.0 in a codebase which uses `mypy --strict`.

This means that custom subclasses of generic types must provide type parameters (`--disallow-any-generics`).
But it isn't clear what the type parameters are or mean in the 2.0 documentation branch.

### Databases / Backends / Drivers targeted

The target is support for `mypy --strict`, but most particularly the `mypy --disallow-any-generics` setting.

Other type checkers may allow for similar checks, though I'm not aware of the various options.

### Example Use

A long time ago, I included the [example of defining a `utcnow` function](https://docs.sqlalchemy.org/en/latest/core/compiler.html#utc-timestamp-function):
```python
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql import expression
from sqlalchemy.types import DateTime

class utcnow(expression.FunctionElement):
    inherit_cache = True
    type = DateTime()

@compiles(utcnow, "postgresql")
def _pg_utcnow(element: t.Any, compiler: t.Any, **kw: t.Any) -> str:
    return "TIMEZONE('utc', CURRENT_TIMESTAMP)"
```

`FunctionElement` is a generic type. So this should be written

```python
class utcnow(expression.FunctionElement[SOME_TYPE_HERE]):
```

And the documentation should make it clear what that type parameter means to a type checker.
Similar usage holds for all other generics.

Right now, in my working branch I've taken a _guess_ that the type parameter for FunctionElement is the type produced in python, and I've annotated as
```python
class utcnow(expression.FunctionElement[datetime.datetime]):
```
but I want to emphasize that this is a guess.

### Additional context

I'm happy to make incremental doc contributions towards this, but I'm not sure where or how such documentation should go.

**Comment by zzzeek** [comment]
yes that's the correct type, when you invoke ``utcnow()`` and get a result back from the ``Result``, the data will be a Python ``datetime`` object, so that's the type.

**Comment by sirosen** [comment]
Awesome, thanks! It's great to have my particular case confirmed!

Where should I propose doc for this? As a user trying to figure this out, I looked at `FunctionElement` itself, as well as the relevant example. But perhaps there's another suitable place, like the migration guide.

**Comment by zzzeek** [comment]
I think for now we should at least get it into the examples we have , which are for example https://docs.sqlalchemy.org/en/20/core/functions.html#sqlalchemy.sql.functions.GenericFunction .   I dont know how much value a migration guide note would add because most people don't have this use case and those who want to add typing to their functions would likely go and consult current 2.0 docs.

**Comment by zzzeek** [comment]
there's also maybe a general statement on "how are column expressions typed?"  that might be good somewhere, but I'm not sure where.    we dont really have a chapter on "typing" that's on its own right now.

**Comment by CaselIT** [comment]
>  we dont really have a chapter on "typing" that's on its own right now.

Maybe we should? Mentioning that in general sqlalchemy objectis are generic in the python type etc. Something like this?

Also maybe we could hsve more meaningful generic type parameters names, so for example instead of the classic `T` something like `PyType`?

**Comment by sirosen** [comment]
I just worked something up for the `FunctionElement`/`GenericFunction` case in particular (PR momentarily), and I found the wording quite difficult because I'm trying to add a doc note "this is a `typing.Generic`" without bringing in a huge amount of context about type annotations.

I would be concerned that sprinkling notes on type annotations across the docs might make `sqlalchemy` more difficult for new users.

**Comment by LeonxLJX** [comment]
I'd like to take this one.

Reading the issue, this looks like a documentation-clarity gap rather than a code defect: the current page doesn't give readers enough to self-diagnose, which is why it surfaces here as an issue instead of being answered by the docs.

Plan: locate the relevant section, confirm what the code actually does today (so the docs match behavior, not intent), tighten the wording, and scan neighboring pages for the same inconsistency — otherwise the fix just relocates the confusion.

One thing to confirm before writing: minimal correction scoped to this page, or also cross-link from related guides if the same gap shows up there?

(claiming via @LeonxLJX)
